src/k8s/pkg/k8sd/api/kubeconfig.go

```diff
@@ -1,25 +1,39 @@
 package api
 
 import (
+	"encoding/json"
 	"fmt"
 	"net/http"
-	"os"
 
 	apiv1 "github.com/canonical/k8s/api/v1"
 	"github.com/canonical/lxd/lxd/response"
 	"github.com/canonical/microcluster/state"
+	"k8s.io/client-go/tools/clientcmd"
 )
 
 func getKubeconfig(s *state.State, r *http.Request) response.Response {
-	// TODO: Render a new kubeconfig instead of reading the existing one
-	//       when the config can be altered via request parameters.
```

Comment:
I'm not "rendering a new kubeconfig" here, but i am returning one that has been modified by request params.  I think that covers the spirit of the above comment.

Reply:
So, here we read in `/etc/kubernetes/admin.conf` (which might have been edited by a user already, and we do not know that.

Starting from the fact that a proper kubeconfig file looks like this:

```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURHekNDQWdPZ0F3SUJBZ0lRQXN4WUpvdW45UU5lQmNseUdVMUdNekFOQmdrcWhraUc5dzBCQVFzRkFEQVkKTVJZd0ZBWURWUVFERXcxcmRXSmxjbTVsZEdWekxXTmhNQjRYRFRJME1ESXlPREUyTWpJeU9Gb1hEVE0wTURJeQpPREUyTWpJeU9Gb3dHREVXTUJRR0ExVUVBeE1OYTNWaVpYSnVaWFJsY3kxallUQ0NBU0l3RFFZSktvWklodmNOCkFRRUJCUUFEZ2dFUEFEQ0NBUW9DZ2dFQkFPbEJFQjRmcHFveHBUbDNnK0JNUHNmSWhtVmk4YUpLcFozdU5vYjAKMkI5TEVvQkw5REVNRG9PdVJCbXRsTmExdEFQaXdTclNKN09TaGluaW5aTmZlQTNmQnAwaHZqQ3BhMWg4MytPRQpYYncyZDNlMFF2Z20wWDIyVEs2THFmZUdBd1BkVXJjQ0RMNVk5UTl3QkoxTE5NdUU4eVNNSnZVS0lyMGdJMXdKCmsvVnFvcWdqMWtKQ0NMTldYeWdyQk5DQmJXQlFtaDgrK0Vsd0NjMW85NnNyM0JaOTh0OHBxTXFBMDRjSmYyYjAKM1pQM24yTk9JYVFYaktXbUhMRWo5elV3eDkrM0JqdEZCeldEcTVVQXl6NDg2a0ZpalJsbjRPU0JhMGk3M0VWaQpJSyszK1M3cU1kSkVzL0lJZGlUcHZDM0JHVnE1UVkvTkdJZnBjdXFoR3dvUUlVc0NBd0VBQWFOaE1GOHdEZ1lEClZSMFBBUUgvQkFRREFnS0VNQjBHQTFVZEpRUVdNQlFHQ0NzR0FRVUZCd01DQmdnckJnRUZCUWNEQVRBUEJnTlYKSFJNQkFmOEVCVEFEQVFIL01CMEdBMVVkRGdRV0JCU2lQeWNwUmxJU2w3QmM0S3E1LzdnazFkcElzVEFOQmdrcQpoa2lHOXcwQkFRc0ZBQU9DQVFFQUk5OEo3U2pZR0VCZ2F4dk01OStaYTY2dnVIY2RUY1o3SXo1ZmdGQTJ5QksrCkpEVU9jZEo2VnB3UmlXK0l4eUQ4OCtuL29kNWJqSGt4ZE1kbm9ZOUNZOXcrRHJWL0FwK09rUWRvNDJJZmNQMGkKWTA3UEFiZS9FVFI3cGFxOFNsUGpxTXV0ZERwanRZTVVMdk5HRkpGbll5SXZjTHA1VDdQazRITmsyOEtXWDhUSAp1ei9QR3BZdXRJSGM4NnV1bVZTV3RnM2craUhzdUgzOFQzaVF0RUs4b2dTMEhCRk0yTy8xbzEwZnMycHlkM3QwCm1aTGdyV3VSS3poTXl5a20vVlFTTHpFSE5CbkZUUW5tSXhxLzRJVUoraVd3SmVtVGxTK0twUFBLc2srTHlNUjEKVGhIWUV1d0NlM2czcWt0M2NnYWJCL1ZpRnlKYU1pdkppSVNsTmdaSmZ3PT0KLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo=
    server: https://127.0.0.1:6443
  name: k8s
contexts:
- context:
    cluster: k8s
    user: k8s-user
  name: k8s
current-context: k8s
kind: Config
users:
- name: k8s-user
  user:
    token: token::e652c6433bf2aaf16afa0140928f21737bb5dabb
```

The three pieces of information we need are the following:

- Cluster CA, which we can retrieve through the ClusterConfig (by means of https://github.com/canonical/k8s-snap/blob/460c50f7905765911298c850519f3238443c2ff5/src/k8s/pkg/k8sd/api/component.go#L86-L89)
- Server URL, which we have through the node address, or the parameter. Note that, if parameter is not set, we must hardcode 6443 here, but rather use the SecurePort from the cluster config https://github.com/canonical/k8s-snap/blob/a6e1a8bded087ed8006b3dee3d6470af92e5791a/src/k8s/pkg/k8sd/types/cluster_config.go#L44)
- Admin token, which we can retrieve by means of https://github.com/canonical/k8s-snap/blob/73c17ef31708078874ba4a45f5c599a705c3458e/src/k8s/pkg/k8sd/api/kubernetes_auth_tokens.go#L40-L44

We should not be reading in a file here, but rather:
- retrieve the info above
- build a new `*clientcmdapi.Config` object with this info (the same type that is returned by `clientcmd.GetConfigFromFileOrDie()`
- render it (as we do now) and return to the user.

Reply:
ack.  big thanks for the pointers to kubeconfig components and walkthrough.  i initially thought it better to read/mod `admin.conf` vs rendering one again, *especially* since it may have been tweaked by the user.

but knowing how many times i've broken kubeconfigs with typos, i see the point in having k8s render something with the things it knows should work.